src/terminal/input/terminalInput.cpp

```diff
@@ -374,25 +374,16 @@ bool TerminalInput::HandleKey(const IInputEvent* const pInEvent) const
         // Only need to handle key down. See raw key handler (see RawReadWaitRoutine in stream.cpp)
         if (keyEvent.IsKeyDown())
         {
-            // For AltGr enabled keyboards, the Windows system will
-            // emit Left Ctrl + Right Alt as the modifier keys and
-            // will have pretranslated the UnicodeChar to the proper
-            // alternative value.
-            // Through testing with Ubuntu, PuTTY, and Emacs for
-            // Windows, it was discovered that any instance of Left
-            // Ctrl + Right Alt will strip out those two modifiers and
-            // send the unicode value straight through to the system.
-            // Holding additional modifiers in addition to Left Ctrl +
-            // Right Alt will then light those modifiers up again for
-            // the unicode value.
-            // Therefore to handle AltGr properly, our first step
-            // needs to be to check if both Left Ctrl + Right Alt are
-            // pressed...
-            // ... and if they are both pressed, strip them out of the control key state.
+            // AltGr key combinations don't always contain any meaningful,
+            // pretranslated unicode character during WM_KEYDOWN.
+            // E.g. on a German keyboard AltGr+Q should result in the "@" character,
+            // but actually results in "Q" with Alt and Ctrl modifier states.
+            // By returning false though, we can abort handling this WM_KEYDOWN
+            // event and let the WM_CHAR handler kick in, which will be
+            // provided with an appropriate unicode character.
             if (keyEvent.IsAltGrPressed())
             {
-                keyEvent.DeactivateModifierKey(ModifierKeyState::LeftCtrl);
-                keyEvent.DeactivateModifierKey(ModifierKeyState::RightAlt);
+                return false;

```

Comment:
Regarding the upcoming review:
I'm particularly concerned about this part of the code, as I'm uncertain which non-Cascadia parts of this project might be negatively affected.
Do you think I should rather move this check into Cascadia (e.g. into `Terminal::SendKeyEvent` above)?

Reply:
Honestly I kinda feel like `terminalInput.cpp` should actually remain untouched and the following should be changed to properly "pretranslate the UnicodeChar to the proper alternative value" instead (like the original comment claims):
https://github.com/microsoft/terminal/blob/bc236c7c59b3b19765e20e361b0039c2d12b81f8/src/cascadia/TerminalCore/Terminal.cpp#L221-L237

Reply:
In fact it seems like the "old" Terminal version did just that if I'm not mistaken:
https://github.com/microsoft/terminal/blob/b9e66fee6d6e131b6711b8e44607d16584d05c98/src/interactivity/win32/windowio.cpp#L201-L205

Reply:
Let me look into this.

Reply:
Yes, correct. Please do move this concern into the `TerminalCore/Terminal.cpp` method of `Terminal::SendKeyEvent` if possible to resolve this issue.

Modifying the `terminal/input/terminalInput.cpp` will influence the behavior of client applications hosted in the ConPTY that are submitting input information with the `WriteConsoleInput` method.

Reply:
I've reverted the entire file. 🙂